test/e2e/framework/service_util.go

```diff
@@ -1450,3 +1461,50 @@ func GetServiceLoadBalancerCreationTimeout(cs clientset.Interface) time.Duration
 	}
 	return LoadBalancerCreateTimeoutDefault
 }
+
+func TestAffinity(jig *ServiceTestJig, execPod *v1.Pod, targetIp string, targetPort int, shouldHold bool) bool {
+	var hostname, response string
+	var err error
+	targetIpPort := net.JoinHostPort(targetIp, strconv.Itoa(targetPort))
+	cmd := fmt.Sprintf(`wget -qO- http://%s/ -T 2`, targetIpPort)
+	affinityCount := 0
+	timeout := ServiceTestTimeout
+	if execPod == nil {
+		timeout = LoadBalancerPollTimeout
+	}
+	if pollErr := wait.PollImmediate(Poll, timeout, func() (bool, error) {
+		if execPod != nil {
+			response, err = RunHostCmd(execPod.Namespace, execPod.Name, cmd)
+		} else {
+			rawResponse := jig.GetHTTPContent(targetIp, targetPort, timeout, "")
+			response = rawResponse.String()
+		}
+		if err != nil || len(response) == 0 {
+			Logf("Failed to get response from %s. Retry until timeout", targetIpPort)
+			hostname = ""
+			affinityCount = 0
+			return false, nil
+		}
+		Logf("Got pod response: %s", response)
+		if shouldHold {
+			if hostname == response {
+				affinityCount++
+			} else {
+				affinityCount = 0
```

Comment:
Why do we need to set this to 0 again here?
Should probably do if hostname == "" { hostname = response } else { error }

Reply:
The session affinity configuration takes some time to propagate. The response sequence with a interval of 2 sec normally looks like:
```
response: service-a
response: service-c
response: service-b
response: service-a
response: service-b
response: service-a
response: service-a
response: service-a
response: service-a
response: service-a
response: service-a
response: service-a
...
```
I assume the affinity holds when 15 same response observed in a row.
I am not sure why hostname != "" means an instant error.

Reply:
Are you creating a new service for each test case or testing flipping affinity on/off?

Reply:
I am not creating new services every time since I want to make sure session affinity can be modified after service created.
Test logic added. Now the test will test SessionAffinity=disabled -> enabled -> disabled.